### PR TITLE
Fix : memory leak when switching between fragments using the bottom navigation view

### PR DESCRIPTION
**Pull request by mejdoo**
### Issues

I got this crash while disabling the CrashController class and trying to login.

java.lang.IllegalStateException: Fragment ConversationListManagerFragment{2f98b00 (9dd6363e-40a9-495b-bc37-aedd75a86369)} has not been attached yet.
at androidx.fragment.app.Fragment.getChildFragmentManager(Fragment.java:923)
at com.waz.zclient.utils.extensions.FragmentUtils.getTopMostFragment(FragmentExtensions.kt:16)

### Causes

I tried to follow up and I discovered a memory leak in ConversationListManagerFragment , specifically in the function onNavigationItemSelected. For every selection done in the BottomNavigationView a new instance of ConversationListFragment was created without being used later in many cases. 

### Solutions

I optimised the fragment replacement mechanism in replaceConversationFragment method and I released some observers watching on the UI thread.

Testing

Disable the CrashController
Logout and login to your account

#### APK
[Download build #325](http://10.10.124.11:8080/job/Pull%20Request%20Builder/325/artifact/build/artifact/wire-dev-PR2407-325.apk)
[Download build #326](http://10.10.124.11:8080/job/Pull%20Request%20Builder/326/artifact/build/artifact/wire-dev-PR2407-326.apk)
[Download build #327](http://10.10.124.11:8080/job/Pull%20Request%20Builder/327/artifact/build/artifact/wire-dev-PR2407-327.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 1]
```diff
@@ -483,15 +507,31 @@ class ConversationListManagerFragment extends Fragment
...
       getChildFragmentManager.beginTransaction
         .replace(R.id.fl__conversation_list_main, fragment, tag)
         .addToBackStack(tag)
         .commit
       setConversationListType(listType)
+
+    }
+    if (navController.getCurrentLeftPage != Page.START) {
+      navController.setLeftPage(Page.CONVERSATION_LIST, Tag)
     }
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: duplicate w/ lines 536-538

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 2]
```diff
@@ -483,15 +507,31 @@ class ConversationListManagerFragment extends Fragment
...
   }
 
-  private def replaceConversationFragment(newInstance: Fragment, tag: String, @ConversationListType listType: Int): Unit = {
-    val currentScreen = FragmentUtils.getTopMostFragment(this)
-    if (currentScreen == null || currentScreen.getClass != newInstance.getClass) {
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: I don't see how removing this solved our problem. can you elaborate?

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 3]
```diff
@@ -483,15 +507,31 @@ class ConversationListManagerFragment extends Fragment
...
-      val fragment = Option(getChildFragmentManager.findFragmentByTag(tag)).getOrElse(newInstance)
+  private def replaceConversationFragment(tag: String, @ConversationListType listType: Int): Unit = {
+    val fragment = getChildFragmentManager.findFragmentByTag(tag)
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: I believe `Option(getChildFragmentManager.findFragmentByTag(tag)).getOrElse(newInstance)`
and this logic that you refactored are basically the same. Can you check with Maciek on this usage of Scala?

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 4]
```diff
@@ -128,72 +145,81 @@ class ConversationListManagerFragment extends Fragment
...
-          stripToConversationList()
-          animateOnIncomingCall()
+      selectDefaultConversationType()
+    }
 
-        case _ => //
+    (for {
+      z <- inject[Signal[ZMessaging]]
+      syncSate <- z.syncRequests.syncState(z.selfUserId, SyncMatchers)
+      animType <- inject[ConversationListController].establishedConversations.map(_.nonEmpty).map {
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: ok, we have to define our coding style

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 5]
```diff
@@ -15,17 +15,34 @@
...
+  * (at your option) any later version.
+  *
+  * This program is distributed in the hope that it will be useful,
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  * GNU General Public License for more details.
+  *
+  * You should have received a copy of the GNU General Public License
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: removed

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 6]
```diff
@@ -87,14 +104,14 @@ class ConversationListManagerFragment extends Fragment
...
-  private var confirmationMenu       : ConfirmationMenu     = _
-  private var bottomNavigationBorder : View                 = _
+  private var listLoadingIndicator: LoadingIndicatorView = _
+  private var mainContainer: FrameLayout = _
+  private var confirmationMenu: ConfirmationMenu = _
+  private var bottomNavigationBorder: View = _
+  private var newFragment: Fragment = _
+  private var bottomNavigation: BottomNavigationView = _
 
   private lazy val bottomNavigationView = returning(view[BottomNavigationView](R.id.fragment_conversation_list_manager_bottom_navigation)) { vh =>
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: I deleted the new one

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 7]
```diff
@@ -128,72 +145,81 @@ class ConversationListManagerFragment extends Fragment
...
+    confirmationMenu = returning(findById[ConfirmationMenu](view, R.id.cm__confirm_action_light)) { v =>
+      v.setVisible(false)
+      v.resetFullScreenPadding()
+    }
 
-      bottomNavigationView
+    bottomNavigationView
 
-      bottomNavigationBorder = findById(view, R.id.fragment_conversation_list_manager_view_bottom_border)
+    bottomNavigation = findById(view, R.id.fragment_conversation_list_manager_bottom_navigation)
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: I deleted the new one

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 8]
```diff
@@ -87,14 +104,14 @@ class ConversationListManagerFragment extends Fragment
...
+  private var confirmationMenu: ConfirmationMenu = _
+  private var bottomNavigationBorder: View = _
+  private var newFragment: Fragment = _
+  private var bottomNavigation: BottomNavigationView = _
 
   private lazy val bottomNavigationView = returning(view[BottomNavigationView](R.id.fragment_conversation_list_manager_bottom_navigation)) { vh =>
-    vh.foreach { _.setOnNavigationItemSelectedListener(ConversationListManagerFragment.this)}
-
     convListController.hasConversationsAndArchive.onUi { case (_, hasArchive) =>
       vh.foreach(view => BottomNavigationUtil.setItemVisible(view, R.id.navigation_archive, hasArchive))
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: I didn't delete it

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala [hunk 9]
```diff
@@ -128,72 +145,81 @@ class ConversationListManagerFragment extends Fragment
...
+        stripToConversationList()
+
+      case ConversationChangeRequester.INCOMING_CALL =>
+        stripToConversationList()
+        animateOnIncomingCall()
+
+      case _ => //
+    }
+
+    inject[AccentColorController].accentColor.map(_.color).onUi { c =>
```
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: done

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala (hunk 1 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: removed

File: app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala (hunk 2 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: You are creating a new instance of ConversationListFragment.newNormalInstance() or ConversationListFragment.newFoldersInstance() for every selection in the BottomNavigationView. Then you are comparing currentScreen.getClass/ newInstance.getClass. Finally you are replacing the current fragment with either the new instance or the fragment stored in the backstack.

Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/conversationlist/ConversationListManagerFragment.scala]: You are creating a new instance of ConversationListFragment.newNormalInstance() or ConversationListFragment.newFoldersInstance() for every selection in the BottomNavigationView. Then you are comparing currentScreen.getClass/ newInstance.getClass. Finally you are replacing the current fragment with either the new instance or the fragment stored in the backstack. For the BottomNavigationView, It's recommended to create new instances for the fragments only once and then you can fetch them from the backstack.

**Pull request merged**